002-tidyr-stringr.Rmd

```diff
@@ -0,0 +1,133 @@
+
+Additionally there are a few gaps in the current API:
+
+-   Neither stringr nor tidyr provide a particularly nice way to work with named capture groups.
+    There's also scope for more directly matching column names to patterns if `extract()` didn't attempt to match the syntax of `separate()` so closely, e.g. `str_extract(x, c(type = ".*", "|", fruit = "orange|apple"))`.
+
+-   tidyr only provides one function for splitting a string across rows, rather than columns: `separate_rows()`.
+    What other functions should exist in this family?
+
+-   `tidyr::separate()` can split up a string by position or by pattern.
+    This now seems too clever, and makes it hard to notice that there are two different tools here.
+
+### Components
+
+To find a solution to this problem we need to first enumerate all the parts.
+I think we can break the problem down into two somewhat independent components: how you break the string into pieces, and whether the number of pieces is fixed (and named) or varying (and named).
+In total, I think this gives five ways to split up a string:
+
+-   Into a fixed number of pieces by length/position, e.g. `"1m15"` -> `c(trt = "1", gender = "m", age = "15")`.
+    This is closely related to `readr::read_fwf()`.
+
+-   Into a varying number of pieces by length, e.g. `"abcdef"` -> `c("ab", "cd", "ef")`
+
+-   Into a fixed number of pieces by a pattern, e.g. `"a/b/c"` -> `c(treatment = "a", group = "b", subgroup = "c").`
+
+-   Into a varying number of pieces by a pattern, e.g. `"a/b/c"` -> `c("a", "b", "c").`
+
+-   Into named groups, e.g. `"a(b)c"` -> `c(treatment = "a", group = "b", subgroup = "c")`
+
+Matching these operations up with the existing stringr and tidyr function gives the following table:
+
+| Split by | Length  | stringr         | tidyr             |
+|----------|---------|-----------------|-------------------|
+| Length   | Fixed   |                 | `separate()`      |
+| Length   | Varying |                 |                   |
+| Pattern  | Fixed   | `str_split_n()` | `separate()`      |
+| Pattern  | Varying | `str_split()`   | `separate_rows()` |
+| Group    | Fixed   | `str_match()`   | `extract()`       |
+
+Note that `separate()` is repeated twice, and there are a number of empty cells.
+Interestingly, there's no stringr function that allows you to extract multiple pieces by position; you currently have to call `str_sub()` multiple times yourself.
+
+Generally, when the pieces might vary from entry to entry you'll want to create rows; when the pieces are fixed, you'll want to create columns.
+This might not be true in every case, but I think it's true in enough cases that we can assume it as a constraint for this problem.
+(Please let me know if you have a counter-example.)
```

Comment:
I agree that when the numbers of pieces might vary from entry to entry it's reasonable to assume that you'll want to create rows. I'm not sure I as strongly agree that when they're fixed you'll want to create columns.

Suppose I have the following data frame:

```r
id value   
 1 "10,20,15"
 2 "20,10,12"
 3 "18,15,11"
```

I  think I'm equally likely to want

```r
id value_1 value_2 value_3
 1 10      20      15     
 2 20      10      12     
 3 18      15      11  
```

as I am to want

```r
id value value_no
 1 10           1
 1 20           2
 1 15           3
 2 20           1
 2 10           2
 2 12           3
 3 18           1
 3 15           2
 3 11           3
```

as a result.

This makes me think I should have the option to use something like `separate_wider()` or `separate_longer()` as opposed to only `separate_auto()` (à la `unnest_auto()`) that will make the decision for me.

Reply:
It's something like: this case is just a special case of the "varying number of pieces" problem where the pieces all have the same number.

Reply:
Putting fixed number of pieces into rows is easy; I'm more concerned about the case where you might want to put varying number of pieces in to columns.

Reply:
Right, I imagine you'll need to define a default value like `NA`?